File: ABC/351-400/355/e.py
N, L, R = map(int, input().split())
ans = []
while L != R:
    i = 0
    while L % pow(2, i+1) == 0 and L+pow(2, i+1) <= R:
        i += 1
    ans.append([i, (pow(2, i)+1)//pow(2,i)])
    L += pow(2, i)

r = 0

for l, r in ans:
    print(f"? {l} {r}")
    x = int(input())
    r += x
    r %= 100

print(f"! {r}")

File: ABC/351-400/355/e.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for(int i=0;i<n;i++)
using ll = long long;
using P = pair<int,int>;

ll pow(int a, int b){
    ll res = 1;
    rep(i,b) res *= a;
    return res;
}

int main(){
    int N,L,R;
    cin >> N >> L >> R;
    vector<P> ans;
    while(L != R){
        int i = 0;
        while(L % pow(2,i+1) == 0 && L+pow(2,i+1) <= R) i++;
        ans.push_back(P(i,L+pow(2,i)));
        L += pow(2,i);
    }
    sort(ans.begin(),ans.end());
    int r = 0;
    for(auto[l,r] : ans){
        cout << "? " << r << ' ' << r << endl;
        int x;
        cin >> x;
        r += x;
        r %= 100;
    }
    cout << "! " << r % 100 << endl;
    return 0;
}